### ISSUE TITLE
Find a way to revert 5ad2aae24474bb5366e84bfe27a5052185d856ce on 16.x

### ISSUE THREAD

**Issue by hvdijk** [llvm/llvm-project#62194]
*sigh*

Yet again communication has broken down.

See discussion at https://reviews.llvm.org/rG5ad2aae24474bb5366e84bfe27a5052185d856ce. This commit, marked NFC, broke code that worked with LLD 14, works with GNU binutils, works with GNU gold.

Its author is unwilling to fix this and in an attempt to get clarification on the current state actually flat out says now that it is unreasonable for the FUSE 2 project to try to maintain binary compatibility with older releases and should just break old binaries.

Given this reasoning, I see no point in continuing to assume good faith and working there to come to an agreeable result, and have to ask you yet again to step in and just revert this broken mess on 16.x, and newer releases until this situation is resolved. Sadly newer changes have since gone in, meaning it does not revert cleanly.

**Comment by MaskRay** [comment]
tl;dr I don't think this commit should be reverted. The release branch should not do anything with it.

People can read my comments on https://reviews.llvm.org/rG5ad2aae24474bb5366e84bfe27a5052185d856ce .
As a symbol versioning expert and a contributor to versioned symbol lookup in glibc rtld, I think fuse 2.x's symbol versioning usage about `VER_NDX_GLOBAL VERSYM_HIDDEN` is unreasonable and ld.lld's behavior is totally fine.

https://maskray.me/blog/2020-11-26-all-about-symbol-versioning#rtld-behavior "it can be bound to foo of non-default version index 2 in relocation resolving phase (not dlsym/dlvsym). The rule retains compatibility when a shared object becomes versioned."

If you remove `fuse_new` from the dynamic symbol table and retain `fuse_new@FUSE_2.2`, `fuse_new@FUSE_2.5`, and `fuse_new@@FUSE_2.6`, an unversioned `fuse_new` will be bound to `fuse_new@FUSE_2.2` (let's say the version `FUSE_2.2` has index 2). This is how compatibility with unversioned shared objects is achieved.

If fuse 2.x wants to retain compatibility with 2.1 symbols, it needs to define another version (used as baseline) `FUSE_2.1`, instead of using the `FUSE_SYMVER(".symver fuse_new_compat2,fuse_new@");` hack (which has unintended outcome; lld's old behavior probably made this work in some cases, but I don't think it is guaranteed).

**Comment by hvdijk** [comment]
@MaskRay You flat out said that wanting to maintain binary compatibility was unreasonable. That is what I opened this issue on, if that's the view you take, any further communication is useless, and you sure as hell removed any willingness on my part to continue interacting with you on this.

**Comment by MaskRay** [comment]
I'm sorry to say that I feel frustrated about some of your comments.
We should be careful in the words that we choose and try to use precise terms to avoid communication problems.

I think I was very careful in how I communicated on https://reviews.llvm.org/rG5ad2aae24474bb5366e84bfe27a5052185d856ce .
Thankfully, I haven't edited any comments, so people can verify. I even asked the binutils folks about their opinions, although there has been no answer so far.

In https://reviews.llvm.org/rG5ad2aae24474bb5366e84bfe27a5052185d856ce, my comments did not say anything about "compatible."
I am focusing on whether the symbol versioning usage in the fuse 2 shared objects is reasonable. (I did not say that
the fuse 2 case cannot be repaired.)
My conclusion is that it is not (https://sourceware.org/pipermail/binutils/2022-November/124312.html).

It's even stranger that they have `fuse_new` (`VER_NDX_GLOBAL VERSYM_HIDDEN`) aliasing `fuse_new_compat2@@FUSE_2.2` (why export a symbol named `*_compat*` at an old version?)

---

Okay, I understand that you felt frustrated when you didn't get what you wanted, but please be mindful of how others would perceive your communication.

I am still unsure how lld's new behavior breaks your objects linking against fuse 2 on your system.
If the problem is the following difference (https://reviews.llvm.org/rG5ad2aae24474bb5366e84bfe27a5052185d856ce#1142446):
```
$ cat test.c
int main() { fuse_new(); }
$ clang test.c -w -s -fuse-ld=bfd -lfuse && nm -D a.out | grep fuse_new
         U fuse_new@FUSE_2.6
$ clang test.c -w -s -fuse-ld=gold -lfuse && nm -D a.out | grep fuse_new
         U fuse_new@FUSE_2.6
$ clang test.c -w -s -fuse-ld=lld -lfuse && nm -D a.out | grep fuse_new
         U fuse_new
```

As I mentioned, ensure libfuse2.so doesn't have the `fuse_new` (`VER_NDX_GLOBAL VERSYM_HIDDEN`) symbol.
Removing this `VER_NDX_GLOBAL` symbol doesn't break compatibility for unversioned `fuse_new` references.
rtld binds such unversioned references to a non-default version definition of index 2.
(See the https://maskray.me/blog/2020-11-26-all-about-symbol-versioning#rtld-behavior section referencd by https://reviews.llvm.org/rG5ad2aae24474bb5366e84bfe27a5052185d856ce#1195621)

It is very likely your system doesn't have any dependent shared objects having unversioned `fuse_new` references.
Those are prebuilt shared objects before libfuse2 started to use `-Wl,--version-script=` (circa 2004).

(
Just FYI for what they should have done for proper symbol versioning.

From libfuse2.so I just built, it seems that fuse 2 uses `fuse_new` to indicate something like `fuse_new@FUSE_2.1`, since the address is different from that of `fuse_new@FUSE_2.2`.
I believe they should define a version node named `FUSE_2.1` and define `fuse_new@FUSE_2.1` (index 2) instead of `fuse_new`.
Unversioned references will bind to `fuse_new@FUSE_2.1`. Your `clang test.c -w -s -fuse-ld=lld -lfuse && nm -D a.out | grep fuse_new` example will get `fuse_new@@FUSE_2.6`.

```
% readelf -W --dyn-syms .libs/libfuse.so.2.9.9 | grep fuse_new
   197: 0000000000013bc0    63 FUNC    GLOBAL DEFAULT   13 fuse_new@FUSE_2.5
   198: 0000000000013b80    16 FUNC    GLOBAL DEFAULT   13 fuse_new
   199: 0000000000013840     8 FUNC    GLOBAL DEFAULT   13 fuse_new@@FUSE_2.6
   200: 0000000000013b70    11 FUNC    GLOBAL DEFAULT   13 fuse_new@FUSE_2.2
   213: 0000000000013b90    35 FUNC    GLOBAL DEFAULT   13 fuse_new_compat1@@FUSE_2.2
   215: 0000000000013b80    16 FUNC    GLOBAL DEFAULT   13 fuse_new_compat2@@FUSE_2.2
   225: 0000000000013b70    11 FUNC    GLOBAL DEFAULT   13 fuse_new_compat22@@FUSE_2.5
   227: 0000000000013bc0    63 FUNC    GLOBAL DEFAULT   13 fuse_new_compat25@@FUSE_2.6
```
)

**Comment by hvdijk** [comment]
> In https://reviews.llvm.org/rG5ad2aae24474bb5366e84bfe27a5052185d856ce, my comments did not say anything about "compatible."
> I am focusing on whether the symbol versioning usage in the fuse 2 shared objects is reasonable. (I did not say that
the fuse 2 case cannot be repaired.)

Your exact quote was "What FUSE 2 intends to do is not reasonable." What FUSE 2 intends to do isn't do trickery with symbol versioning, it's maintain binary compatibility. You may feel that what they are doing to achieve that is not reasonable, but that's not what you said, you said that that itself wasn't reasonable, after a comment of mine in which I had already drawn a distinction between the two when I asked how they should be going about achieving that instead. Based on what you said, I feel that my reaction of wishing to cease engagement is entirely appropriate. Are you now saying that this is basically the exact opposite of what you actually meant?

**Comment by MaskRay** [comment]
libfuse 2's symbol versioning is trickery.

```
% grep fuse_new@ lib/fuse.c
FUSE_SYMVER(".symver fuse_new_compat2,fuse_new@");           /// unreasonable
FUSE_SYMVER(".symver fuse_new_compat22,fuse_new@FUSE_2.2");  /// correct
FUSE_SYMVER(".symver fuse_new_compat25,fuse_new@FUSE_2.5");  /// correct
% readelf -W --dyn-syms lib/.libs/libfuse.so.2.9.9 | grep fuse_new
   126: 000000000001d380     8 FUNC    GLOBAL DEFAULT   13 fuse_new@@FUSE_2.6
   127: 000000000001d6c0    16 FUNC    GLOBAL DEFAULT   13 fuse_new
   128: 000000000001d6b0    11 FUNC    GLOBAL DEFAULT   13 fuse_new@FUSE_2.2
   129: 000000000001d700    63 FUNC    GLOBAL DEFAULT   13 fuse_new@FUSE_2.5
   138: 000000000001d6d0    35 FUNC    GLOBAL DEFAULT   13 fuse_new_compat1@@FUSE_2.2
   144: 000000000001d6c0    16 FUNC    GLOBAL DEFAULT   13 fuse_new_compat2@@FUSE_2.2
   277: 000000000001d6b0    11 FUNC    GLOBAL DEFAULT   13 fuse_new_compat22@@FUSE_2.5
   287: 000000000001d700    63 FUNC    GLOBAL DEFAULT   13 fuse_new_compat25@@FUSE_2.6
```

In the built shared object, version index 2 (the first version definition) is `FUSE_2.2`.

If libfuse 2 wants to provide compatibility with a version before `FUSE_2.2`, say, 2.1, it should use

```c
FUSE_SYMVER(".symver fuse_new_compat2,fuse_new@2.1");
```
and add `FUSE_2.1 {};` before `FUSE_2.2` in `lib/fuse_versionscript`.
(It's not necessary to enumerate all `xxx@FUSE_2.2` and place them in `FUSE_2.1 {};`. An empty list suffices.)

If we don't have unversioned `fuse_new` references built before libfuse 2 got versioned (circa 2004?), we can simply remove `FUSE_SYMVER(".symver fuse_new_compat2,fuse_new@");`.

---

libfuse 3's symbol versioning looks correct to me. It provides compatibility for 3.0 (circa 2012), but not for older versions.

```
% grep fuse_new@ lib/fuse.c
FUSE_SYMVER("fuse_new_31", "fuse_new@@FUSE_3.1")
FUSE_SYMVER("fuse_new_30", "fuse_new@FUSE_3.0")
% readelf -W --dyn-syms out/release/lib/libfuse3.so.3.14.1 | grep fuse_new
   250: 000000000001d190   216 FUNC    GLOBAL DEFAULT   13 fuse_new_30@@FUSE_3.1
   254: 000000000001cbc0  1477 FUNC    GLOBAL DEFAULT   13 fuse_new_31@@FUSE_3.1
   274: 000000000001cbc0  1477 FUNC    GLOBAL DEFAULT   13 fuse_new@@FUSE_3.1
   275: 000000000001d190   216 FUNC    GLOBAL DEFAULT   13 fuse_new@FUSE_3.0
```

Note that it uses `fuse_new@FUSE_3.0` instead of `fuse_new@` for the lowest supported version.
(I'd ensure `fuse_new_30` and `fuse_new_31` are removed from the dynamic symbol table by using `local:`.)

**Comment by hvdijk** [comment]
It will be quite a while before I can fully verify that this works as intended, but I see no reason to doubt it. If it does, I agree that we have no reason to revert anything, and so I will pre-emptively close this issue.

I am going to take you at your word that you had meant for your earlier comment to be interpreted the way you did, and apologise for the escalation. I had taken an extended break from LLVM after bad experiences. I have recently restarted trying to contribute, and I suspect this was too soon and I continue to judge based on the past. I will try to stay away from LLVM development for longer so as to avoid similar conflicts.
